### PR TITLE
Make pango markup in mode names optional with a flag.

### LINKED ISSUE #1992: < in mode names does not work
From http://bugs.debian.org/801574:

> Dear Maintainer,
> In 4.10 this line was displayed when the mode was switched on:
>  set $mode_randr xrandr off (o), xrandr right (->), xrandr left (<)
> In 4.11 it does not show, I determined that it is the fault of '<' character. When you delete it, the line show again.

i3 v4.11

### PR DESCRIPTION
**Pull request by Airblader**
This introduces the flag "--pango" on the mode config directive to
explicitly enable pango markup for mode names. Not setting this will
cause the mode name to be rendered as is.
This fixes a regression in 4.11 where mode names containing characters
such as '<' would break user's configs as they didn't escape these
characters.

fixes #1992

### CODE REVIEW

**Comment by Airblader** [comment]
@stapelberg This is quite a bit of code, but we need it to propagate it all the way through. If you want to change your mind and accept the minor breakage, I won't be sad about rejecting the PR. ;)

Either way, I've noticed that the userguide doesn't really document modes except for the resize mode example. We should change that, but in a separate PR, independently of this change.

**Comment by Airblader** [comment]
By the way, the flag could also be implemented on the command, but I didn't really see what the point is in dynamically deciding whether or not to use it, so I put it on the config. I assume that's what the suggestion was anyway.

File: src/config_directives.c [hunk 1]
```diff
@@ -129,6 +130,7 @@ CFGFUN(enter_mode, const char *modename) {
     DLOG("\t now in mode %s\n", modename);
     FREE(current_mode);
     current_mode = sstrdup(modename);
+    current_mode_is_markup = pango != NULL;
```
Comment (stapelberg) [inline src/config_directives.c]: For consistency, please add parens around the expression, i.e. `current_mode_is_markup = (pango != NULL);`

File: src/bindings.c [hunk 2]
```diff
@@ -437,7 +438,7 @@ void switch_mode(const char *new_mode) {
         grab_all_keys(conn);
 
         char *event_msg;
-        sasprintf(&event_msg, "{\"change\":\"%s\"}", mode->name);
+        sasprintf(&event_msg, "{\"change\":\"%s\",\"is_markup\":%s}", mode->name, mode->is_markup ? "true" : "false");
```
Comment (stapelberg) [inline src/bindings.c]: For consistency, please add parens around the ternary if, i.e. `(mode->is_markup ? "true" : "false")`

File: i3bar/src/mode.c [hunk 3]
```diff
@@ -31,17 +33,35 @@ static int mode_string_cb(void *params_, const unsigned char *val, size_t len) {
...
-        /* Save the name */
-        params->mode->name = i3string_from_markup_with_length((const char *)val, len);
-        /* Save its rendered width */
-        params->mode->width = predict_text_width(params->mode->name);
+        char *copy = smalloc(sizeof(const unsigned char) * (len + 1));
+        strncpy(copy, (const char *)val, len);
+        copy[len] = '\0';
 
-        DLOG("Got mode change: %s\n", i3string_as_utf8(params->mode->name));
+        params->name = copy;
```
Comment (stapelberg) [inline i3bar/src/mode.c]: The following would be more in line with how we do string processing in the rest of the codebase, I think:

```c
sasprintf(&(params->name), "%.*s", len, val);
```

**Comment by stapelberg** [comment]
It’s a bit unclean that in one half of the code, the flag is called `pango`, while in the rest, it’s called `is_markup`. Should we rename it to `pango_markup` in all the places, including the flag name itself?

File: i3bar/src/mode.c (hunk 3 above)
Comment (Airblader) [inline i3bar/src/mode.c]: Not in the i3bar code. :) I agree that the other way is prettier, but let's do a separate PR. I'll create a tag for technical improvements and open an issue.

**Comment by Airblader** [comment]
@stapelberg Done. I didn't rename the field in the pre-existing IPC communication as this potentially (though unlikely) can break things for users.

**Pull request merged**